### PR TITLE
Added merkle leaf verification, fixed bugs.

### LINKED ISSUE #6: Verifying that a MerkleTree contains a given leaf
The Merkle tree implementation should be able to prove that it contains a given leaf. It should be able to do this with data-transfer-size complexity O(log n).

### PR DESCRIPTION
**Pull request by mileswatson**
Implements #6

### CODE REVIEW

File: src/crypto/hashing/merkle.rs [hunk 1]
```diff
@@ -81,20 +81,93 @@ impl MerkleTree {
...
+        verify_proof_rec(index, left_size, leaf, &proof[1..]).map(|left| hash![left, proof[0]])
+    } else {
+        let relative_index = index - left_size;
+        verify_proof_rec(relative_index, size - left_size, leaf, &proof[1..])
+            .map(|right| hash![proof[0], right])
+    }
+}
+
+/// Gets the size of the left child, given the size of a parent MerkleNode
+fn left_child_size(mut size: usize) -> usize {
```
Comment (brianlck) [inline src/crypto/hashing/merkle.rs]: left_child_size can be further optimized from logarithmic time to constant time by bit manipulation

File: src/crypto/hashing/merkle.rs [hunk 2]
```diff
@@ -110,12 +183,40 @@ impl Eq for MerkleTree {}
...
         MerkleTree::new::<u8>(&[1, 2]);
         MerkleTree::new::<u8>(&[1, 2, 3]);
         MerkleTree::new::<u8>(&[1, 2, 3, 4]);
         MerkleTree::new::<u8>(&[1, 2, 3, 4, 5]);
     }
+
+    #[test]
+    fn child_size() {
+        assert_eq!(left_child_size(0), 0);
+        assert_eq!(left_child_size(1), 1);
```
Comment (brianlck) [inline src/crypto/hashing/merkle.rs]: is left_child_size(1) really 1?

Comment (brianlck) [reply src/crypto/hashing/merkle.rs]: I believe only a leaf node would have size of 1?

File: src/crypto/hashing/merkle.rs [hunk 3]
```diff
@@ -81,20 +81,93 @@ impl MerkleTree {
...
+
+    /// Generate a proof that the given item is contained within the Merkle tree.
+    pub fn construct_proof(&self, index: usize) -> Vec<Hash> {
+        if index >= self.size {
+            panic!()
+        } else {
+            let mut proof = vec![];
+            let mut relative_index = index;
+            let mut current = &self.nodes[self.root];
+            while current.size > 1 {
```
Comment (brianlck) [inline src/crypto/hashing/merkle.rs:102]: Should there be a check of index < current.size on each iteration?

Comment (brianlck) [reply src/crypto/hashing/merkle.rs:102]: Sorry I meant relative_index not index

File: src/crypto/hashing/merkle.rs (hunk 2 above)
Comment (mileswatson) [reply src/crypto/hashing/merkle.rs]: You are correct. left_child_size(1) should be 0 (or it should panic for size < 2).

File: src/crypto/hashing/merkle.rs (hunk 3 above)
Comment (mileswatson) [reply src/crypto/hashing/merkle.rs:102]: We only need to check once, as long as left_child_size(x) < x for all x.

**Pull request merged**